File: frontend/src/api/management.ts
/** Tipos del panel de gerencia: espejo de backend/app/schemas/management.py.
 *
 * Importes en céntimos; tasas como fracción 0-1, o null cuando no hay
 * denominador (sin clientes no hay "0 %", hay "sin datos"). Fechas como
 * string ISO en hora local del negocio (period.timezone), sin zona. */

export type PeriodDays = 7 | 30 | 90 | 365

export interface PeriodInfo {
  days: number
  start: string
  end: string
  timezone: string
  bucket: "day" | "week"
}

export interface SeriesPoint {
  bucket: string
  signups: number
  active_customers: number
  revenue_cents: number
}

export interface Overview {
  period: PeriodInfo
  customers_total: number
  new_customers: number
  new_customers_prev: number
  active_customers: number
  active_customers_prev: number
  dau: number
  wau: number
  mau: number
  stickiness: number | null
  activation_rate: number | null
  paying_customers: number
  mrr_cents: number
  arppu_cents: number | null
  paid_conversion: number | null
  revenue_cents: number
  revenue_prev_cents: number
  series: SeriesPoint[]
}

export type Segment = "new" | "current" | "at_risk" | "dormant" | "never_active"

export interface Lifecycle {
  segments: Partial<Record<Segment, number>>
  last_week_active: number
  retained_this_week: number
  curr: number | null
}

export interface Engagement {
  period: PeriodInfo
  features: { feature: string; customers: number; volume: number }[]
  tutor: {
    sessions: number
    customers: number
    learner_turns: number
    avg_turns_per_session: number | null
    median_minutes: number | null
    correction_rate: number | null
    library_sessions: number
    module_sessions: number
  }
  heatmap: { weekday: number; hour: number; actions: number; customers: number }[]
  lifecycle: Lifecycle
  top_corrections: { rule: string; times: number; customers: number }[]
  screens: { path: string | null; views: number; customers: number }[]
}

export type FunnelKey = "registered" | "activated" | "engaged" | "achieved" | "paid"

export interface Journey {
  period: PeriodInfo
  funnel: { key: FunnelKey; customers: number }[]
  paid_any: number
  median_hours_to_activate: number | null
  median_days_to_achieve: number | null
  median_days_to_pay: number | null
  modules: { code: string | null; title: string; level_code: string; started: number; completed: number }[]
  outcomes: {
    exams_taken: number
    exams_passed: number
    modules_completed: number
    levels_certified: number
    tutor_tasks: number
    tutor_tasks_completed: number
  }
  levels: { level_code: string; customers: number }[]
}

export interface Retention {
  weeks: number
  cohorts: { week_start: string; size: number; retention: (number | null)[] }[]
  day_n: { day: number; eligible: number; retained: number; rate: number | null }[]
  lifecycle: Lifecycle
}

export interface Revenue {
  period: PeriodInfo
  mrr_cents: number
  paying_customers: number
  arppu_cents: number | null
  revenue_cents: number
  revenue_prev_cents: number
  paid_conversion: number | null
  series: SeriesPoint[]
  by_provider: { provider: string; payments: number; amount_cents: number }[]
  by_status: { status: string; payments: number; amount_cents: number }[]
  subscriptions_by_status: Record<string, number>
  pending_verification: number
  pending_verification_cents: number
  subscribers_at_start: number
  new_subscribers: number
  churned_in_period: number
  churn_rate: number | null
  ltv_cents: number | null
}

export interface CustomerRow {
  id: string
  full_name: string
  email: string
  created_at: string
  level_code: string | null
  last_active_at: string | null
  active_days_30: number
  tutor_messages: number
  modules_completed: number
  subscription_status: string
  subscription_period_end: string | null
  total_paid_cents: number
  segment: Segment
}

export type CustomerSort = "last_active" | "signup" | "messages" | "paid" | "name"

export interface CustomerList {
  items: CustomerRow[]
  total: number
  /** "Ahora" en hora del negocio, referencia para los "hace N días". */
  as_of: string
}

export interface CustomerDetail {
  as_of: string
  customer: CustomerRow
  milestones: {
    first_learning_at: string | null
    first_tutor_at: string | null
    first_module_completed_at: string | null
    first_level_certified_at: string | null
    first_payment_at: string | null
    tutor_sessions: number
    exam_answers: number
    modules_in_progress: number
    active_days_total: number
  }
  daily_activity: { day: string; actions: number }[]
  timeline: { at: string; kind: string; detail: Record<string, unknown> }[]
  top_corrections: { rule: string; times: number }[]
}

// --- Pagos (revisión manual) -------------------------------------------------

export type PaymentStatus = "pending_verification" | "approved" | "rejected" | "failed" | "refunded"

export interface PaymentReviewRow {
  id: string
  provider: "pago_movil" | "binance_pay" | "credit_card" | "paypal"
  status: PaymentStatus
  amount_cents: number
  currency: string
  external_reference: string | null
  /** Lo que declaró el alumno y lo que se le pidió. Pago Móvil:
   * reference_number, payer_bank/cedula/phone, amount_bs, expected_amount_bs,
   * bs_per_usd, rate_value_date. Binance: order_id, payer_account,
   * expected_amount, asset. Rechazado: rejection_reason. */
  payload: Record<string, string | number | null | undefined>
  created_at: string
  reviewed_at: string | null
  reviewer_name: string | null
  customer: { id: string; full_name: string; email: string }
  access_until: string | null
}

export interface PaymentReviewList {
  items: PaymentReviewRow[]
  total: number
  as_of: string
}

export interface PaymentReviewSummary {
  pending: number
  pending_by_provider: Record<string, number>
  oldest_pending_at: string | null
  approved_today: number
  approved_today_cents: number
}

// --- Correos ---------------------------------------------------------------

export type EmailStatus = "sent" | "failed" | "skipped" | "sending"

export interface EmailMessageRow {
  id: string
  /** "bienvenida", "pago_aprobado", "vence_pronto", "campana"... */
  kind: string
  status: EmailStatus
  to_email: string
  subject: string
  error: string | null
  created_at: string
  customer: { id: string; full_name: string; email: string } | null
}

export interface EmailMessageList {
  items: EmailMessageRow[]
  total: number
  as_of: string
}

export interface EmailSummary {
  days: number
  sent: number
  failed: number
  skipped: number
  failure_rate: number | null
  by_kind: { kind: string; sent: number; failed: number; skipped: number }[]
  last_sent_at: string | null
  /** Por qué no está saliendo ningún correo (falta la clave, modo de
   * pruebas). null = todo normal. */
  blocked_reason: string | null
}

/** Los campos que se escriben en el editor. El HTML no se guarda: se arma
 * al enviar con la plantilla de la marca. */
export interface TemplateDraft {
  name: string
  subject: string
  preheader: string
  eyebrow: string
  title: string
  body: string
  button_label: string | null
  button_url: string | null
}

export interface EmailTemplate extends TemplateDraft {
  id: string
  created_at: string
  updated_at: string
  author_name: string | null
  times_sent: number
  last_sent_at: string | null
}

export interface TemplatePreview {
  subject: string
  html: string
  text: string
}

export interface Audience {
  key: string
  label: string
  description: string
  customers: number
}

export interface Campaign {
  id: string
  name: string
  subject: string
  audience: string
  audience_label: string
  status: "enviando" | "terminada"
  recipients: number
  sent: number
  skipped: number
  failed: number
  sender_name: string | null
  created_at: string
  finished_at: string | null
}

File: backend/app/schemas/management.py
"""Respuestas del panel de gerencia (routers/management.py).

Importes en céntimos (int) como en el resto de la facturación; tasas como
fracción 0-1 (float) o None cuando el denominador es cero — un "0 %" sobre
cero clientes afirmaría algo que no se ha medido.
"""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field


class PeriodInfo(BaseModel):
    days: int
    start: datetime
    end: datetime
    timezone: str
    bucket: str  # "day" | "week": granularidad de las series


class SeriesPoint(BaseModel):
    bucket: datetime
    signups: int
    active_customers: int
    revenue_cents: int


class OverviewOut(BaseModel):
    period: PeriodInfo
    customers_total: int
    new_customers: int
    new_customers_prev: int
    active_customers: int
    active_customers_prev: int
    dau: int
    wau: int
    mau: int
    # DAU medio de 30 días / MAU: qué fracción del mes vuelve cada día.
    stickiness: float | None
    # Altas del periodo que practicaron en sus primeros 7 días.
    activation_rate: float | None
    paying_customers: int
    mrr_cents: int
    arppu_cents: int | None
    # Clientes que han pagado alguna vez / clientes totales.
    paid_conversion: float | None
    revenue_cents: int
    revenue_prev_cents: int
    series: list[SeriesPoint]


class FeatureUsage(BaseModel):
    feature: str
    customers: int
    volume: int


class TutorStats(BaseModel):
    sessions: int
    customers: int
    learner_turns: int
    avg_turns_per_session: float | None
    median_minutes: float | None
    # Turnos del alumno en los que el tutor corrigió algo.
    correction_rate: float | None
    library_sessions: int
    module_sessions: int


class HeatCell(BaseModel):
    weekday: int  # 1 = lunes … 7 = domingo
    hour: int  # 0-23, hora local
    actions: int
    customers: int


class Lifecycle(BaseModel):
    segments: dict[str, int]
    last_week_active: int
    retained_this_week: int
    curr: float | None


class CorrectionRule(BaseModel):
    rule: str
    times: int
    customers: int


class ScreenViews(BaseModel):
    path: str | None
    views: int
    customers: int


class EngagementOut(BaseModel):
    period: PeriodInfo
    features: list[FeatureUsage]
    tutor: TutorStats
    heatmap: list[HeatCell]
    lifecycle: Lifecycle
    top_corrections: list[CorrectionRule]
    screens: list[ScreenViews]


class FunnelStage(BaseModel):
    key: str
    customers: int


class ModuleProgress(BaseModel):
    code: str | None
    title: str
    level_code: str
    started: int
    completed: int


class LearningOutcomes(BaseModel):
    exams_taken: int
    exams_passed: int
    modules_completed: int
    levels_certified: int
    tutor_tasks: int
    tutor_tasks_completed: int


class LevelCount(BaseModel):
    level_code: str
    customers: int


class JourneyOut(BaseModel):
    period: PeriodInfo
    funnel: list[FunnelStage]
    paid_any: int
    median_hours_to_activate: float | None
    median_days_to_achieve: float | None
    median_days_to_pay: float | None
    modules: list[ModuleProgress]
    outcomes: LearningOutcomes
    levels: list[LevelCount]


class CohortRow(BaseModel):
    week_start: date
    size: int
    # Índice = semanas desde el alta. None = esa semana todavía no ha pasado.
    retention: list[float | None]


class DayNRetention(BaseModel):
    day: int
    eligible: int
    retained: int
    rate: float | None


class RetentionOut(BaseModel):
    weeks: int
    cohorts: list[CohortRow]
    day_n: list[DayNRetention]
    lifecycle: Lifecycle


class ProviderRevenue(BaseModel):
    provider: str
    payments: int
    amount_cents: int


class PaymentStatusCount(BaseModel):
    status: str
    payments: int
    amount_cents: int


class RevenueOut(BaseModel):
    period: PeriodInfo
    mrr_cents: int
    paying_customers: int
    arppu_cents: int | None
    revenue_cents: int
    revenue_prev_cents: int
    paid_conversion: float | None
    series: list[SeriesPoint]
    by_provider: list[ProviderRevenue]
    by_status: list[PaymentStatusCount]
    subscriptions_by_status: dict[str, int]
    pending_verification: int
    pending_verification_cents: int
    subscribers_at_start: int
    new_subscribers: int
    churned_in_period: int
    churn_rate: float | None
    # Estimación: ARPPU / churn mensual. None si no hay churn medible.
    ltv_cents: int | None


class CustomerRow(BaseModel):
    id: uuid.UUID
    full_name: str
    email: str
    created_at: datetime
    level_code: str | None
    last_active_at: datetime | None
    active_days_30: int
    tutor_messages: int
    modules_completed: int
    subscription_status: str
    subscription_period_end: datetime | None
    total_paid_cents: int
    segment: str


class CustomerListOut(BaseModel):
    items: list[CustomerRow]
    total: int
    # "Ahora" en hora del negocio: para que el frontend calcule "hace 3
    # días" contra la misma referencia que las fechas, no contra la zona
    # horaria del navegador de quien mira.
    as_of: datetime


class CustomerMilestones(BaseModel):
    first_learning_at: datetime | None
    first_tutor_at: datetime | None
    first_module_completed_at: datetime | None
    first_level_certified_at: datetime | None
    first_payment_at: datetime | None
    tutor_sessions: int
    exam_answers: int
    modules_in_progress: int
    active_days_total: int


class DailyActivity(BaseModel):
    day: date
    actions: int


class TimelineItem(BaseModel):
    at: datetime
    kind: str
    detail: dict


class CustomerRuleCount(BaseModel):
    rule: str
    times: int


class CustomerDetailOut(BaseModel):
    as_of: datetime
    customer: CustomerRow
    milestones: CustomerMilestones
    daily_activity: list[DailyActivity]
    timeline: list[TimelineItem]
    top_corrections: list[CustomerRuleCount]


# --- Pagos (revisión manual desde el panel de gerencia) ---------------------


class PaymentCustomer(BaseModel):
    id: uuid.UUID
    full_name: str
    email: str


class PaymentReviewRow(BaseModel):
    id: uuid.UUID
    provider: str
    status: str
    amount_cents: int
    currency: str
    external_reference: str | None
    # Lo que declaró el alumno y lo que se le pidió: referencia, banco,
    # cédula, monto en Bs esperado y declarado, ID de orden de Binance...
    payload: dict
    created_at: datetime
    reviewed_at: datetime | None
    reviewer_name: str | None
    customer: PaymentCustomer
    # Hasta cuándo da acceso, si se aprobó.
    access_until: datetime | None


class PaymentReviewList(BaseModel):
    items: list[PaymentReviewRow]
    total: int
    as_of: datetime  # "ahora" en hora del negocio, para los "hace 2 h"


class PaymentReviewSummary(BaseModel):
    pending: int
    pending_by_provider: dict[str, int]
    oldest_pending_at: datetime | None
    approved_today: int
    approved_today_cents: int


class RejectRequest(BaseModel):
    reason: str = Field(min_length=3, max_length=300)


# --- Correos (registro de envíos, plantillas y campañas) --------------------


class EmailMessageRow(BaseModel):
    id: uuid.UUID
    kind: str
    status: str
    to_email: str
    subject: str
    error: str | None
    created_at: datetime
    customer: PaymentCustomer | None


class EmailMessageList(BaseModel):
    items: list[EmailMessageRow]
    total: int
    as_of: datetime


class EmailKindCount(BaseModel):
    kind: str
    sent: int
    failed: int
    skipped: int


class EmailSummary(BaseModel):
    days: int
    sent: int
    failed: int
    skipped: int
    # Correos que no salieron por un fallo / los que se intentaron. None
    # si no se intentó ninguno: un "0 %" sobre cero envíos no dice nada.
    failure_rate: float | None
    by_kind: list[EmailKindCount]
    last_sent_at: datetime | None
    # Si hace falta configurar algo para que salgan correos, se dice aquí
    # y la pantalla lo avisa arriba en vez de mentir con ceros.
    blocked_reason: str | None


class TemplateIn(BaseModel):
    name: str = Field(min_length=2, max_length=120)
    subject: str = Field(min_length=2, max_length=200)
    preheader: str = Field(default="", max_length=200)
    eyebrow: str = Field(default="Espikin", max_length=40)
    title: str = Field(min_length=2, max_length=200)
    body: str = Field(min_length=2, max_length=4000)
    button_label: str | None = Field(default=None, max_length=60)
    button_url: str | None = Field(default=None, max_length=500)


class TemplateOut(TemplateIn):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    author_name: str | None
    # Cuántas veces se ha enviado y cuándo fue la última.
    times_sent: int
    last_sent_at: datetime | None


class TemplatePreviewOut(BaseModel):
    subject: str
    html: str
    text: str


class AudienceOut(BaseModel):
    key: str
    label: str
    description: str
    customers: int


class CampaignIn(BaseModel):
    template_id: uuid.UUID
    audience: str


class CampaignOut(BaseModel):
    id: uuid.UUID
    name: str
    subject: str
    audience: str
    audience_label: str
    status: str
    recipients: int
    sent: int
    skipped: int
    failed: int
    sender_name: str | None
    created_at: datetime
    finished_at: datetime | None
